### PR TITLE
(test) O3-4611: Add E2E test for the Ward app management workflow

### PR DESCRIPTION
**Pull request by jayg2002**
## Requirements

- [X] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [X] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [X] My work includes tests or is validated by existing tests.

## Summary

This PR is used to cover E2E test cases for the Ward app
## Related Issue

https://openmrs.atlassian.net/browse/O3-4611

### CODE REVIEW

File: packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ const WardPatientCard: React.FC<Props> = ({ children, wardPatient }) => {
   const { patient } = wardPatient;
 
   return (
-    <div className={styles.wardPatientCard}>
+    <div data-testid={patient.uuid} className={styles.wardPatientCard}>
```
Comment (UNCANNY69) [inline packages/esm-ward-app/src/ward-patient-card/ward-patient-card.component.tsx:16]: This change was made to make the selection of the Ward patient card possible

**Comment by UNCANNY69** [comment]
Hi @NethmiRodrigo , in this PR there is a problem for the E2E tests to run smoothly as when the ancestor location is Outpatient clinic the Transferrable locations are not available , what should be done here should i pass the `E2E_LOGIN_DEFAULT_LOCATION_UUID` as a variable rather then env variable ?

**Comment by UNCANNY69** [comment]
I have 2 question based on the above issue:
1) Should the Transferrable locations depend on the ancestor location 
2) In this E2E framework why is the location being passed from the environment variable , this reduces the flexibility as visit's can be started from anywhere ?
Any inputs on these would really help my understanding , thanks @NethmiRodrigo

**Comment by NethmiRodrigo** [comment]
> I have 2 question based on the above issue:
> 
> 1. Should the Transferrable locations depend on the ancestor location
> 2. In this E2E framework why is the location being passed from the environment variable , this reduces the flexibility as visit's can be started from anywhere ?
>    Any inputs on these would really help my understanding , thanks @NethmiRodrigo

I'd defer to @jayasanka-sack

File: e2e/specs/wards-managment.spec.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+test.beforeEach(async ({ api }) => {
+  patient = await generateRandomPatient(api);
+  visit = await startVisit(api, patient.uuid);
+});
+
+test('Adding a patient to the ward then transferring and then discharing', async ({ page }) => {
+  const wardAppPage = new WardAppPage(page);
+
+  await test.step('When I go to the Wards page', async () => {
```
Comment (chibongho) [inline e2e/specs/wards-managment.spec.ts:19]: Do we know the initial login location of the user? This affects which location we open the ward in, and the "transfer" test steps might not work as expected unless we can guarantee the login location to be different.

Comment (UNCANNY69) [reply e2e/specs/wards-managment.spec.ts:19]: Hi @chibongho , just some more clarification , how is the location in the identifier and the login location related and in what situation does the transfer locations come up empty 

Currently :

Login location : Outpatient clinic
Identifier location : Outpatient clinic
image
image

**Comment by UNCANNY69** [comment]
> Hey @UNCANNY69 we're passing the location as an environment variable because it's not unique across different implementations. Each setup might have its own locations, so the idea is that anyone should be able to run the E2E suite against their implementation without changes.
> 
> If you need any additional variables and they also vary between implementations, feel free to define them in the .env file as well.

Sure @jayasanka-sack , thanks for the guidence

File: e2e/specs/wards-managment.spec.ts (hunk 2 above)
Comment (chibongho) [reply e2e/specs/wards-managment.spec.ts:19]: Ok, good. Looks like we always set the login location to "Outpatient Clinic". This is what I was looking for. Maybe just put a comment at the top of the test case about that?

I am not sure what identifier location is , but it doesn't come into play here.

Comment (UNCANNY69) [reply e2e/specs/wards-managment.spec.ts:19]: Can the behaviour in the video below be explained , as to why the patient created in E2E tests a normal patient created have this different locations available ? @chibongho 

https://github.com/user-attachments/assets/760c0115-56bf-4cea-902c-0c0ad999ac67

Comment (chibongho) [reply e2e/specs/wards-managment.spec.ts:19]: That's a good point. Transfers / Admissions should only be available to locations tagged as "admission location"s. (You can see the tags (here)[https://dev3.openmrs.org/openmrs/admin/locations/hierarchy.list]). Being able to admit to "Outpatient Clinic", which is not an admission location, is a bug. 

I think the proper thing to do for non-admission locations would be to hide the "Add patient to ward" button on the side rail. 
image
(A more aggressive thing to do would be to hide the "X admission request. __Manage__" bar, but that means patients who got into unexpected edge cases would be stuck.)

As for this test case itself, I think we should switch the login user's visit location to a different admission location. (like "Ward 1").

Comment (UNCANNY69) [reply e2e/specs/wards-managment.spec.ts:19]: I will resolve the issue for the ticket filed, then move ahead with the test case to add a better coverage including the changes implemented in the ticket @chibongho

File: e2e/specs/wards-managment.spec.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByPlaceholder('Search locations').fill('Inpatient Ward');
+    await page.getByText('Inpatient Ward').click();
+    await page.getByRole('button', { name: 'Save', exact: true }).click();
+  });
+
+  await test.step('And I should see a success message for creating transfer request of the patient', async () => {
+    await expect(page.getByText(/patient transfer request created/i)).toBeVisible();
+  });
+
+  await test.step('Then I should change the location to Inpatient', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:78]:
```suggestion
  await test.step('And when I change the location to "Inpatient Ward"', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByRole('button', { name: 'Manage' }).click();
+    await page.getByRole('button', { name: 'Transfer patient', exact: true }).click();
+  });
+
+  await test.step('And I should see the success message and the patient in the Inpatient Ward', async () => {
+    await expect(page.getByText('Patient admitted successfully')).toBeVisible();
+    await expect(page.getByText(patient.person.display)).toBeVisible();
+  });
+
+  await test.step('Then I should change the location and accept the admin request', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:99]:
```suggestion
  await test.step('And when I discharge the patient', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByTestId(patient.uuid).click();
+    await page.getByLabel('Discharge').click();
+    await page.getByRole('button', { name: 'Proceed with patient discharge' }).click();
+  });
+});
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:104]:
```suggestion
  await test.step('Then I should see a success message confirming the patient was discharged', async () => {
    await expect(page.getByText(/patient was  discharged/i)).toBeVisible();
  });
});
```

File: e2e/specs/wards-managment.spec.ts (hunk 2 above)
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:19]:
```suggestion
  await test.step('When I visit the Ward app', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+test('Adding a patient to the ward then transferring and then discharing', async ({ page }) => {
+  const wardAppPage = new WardAppPage(page);
+
+  await test.step('When I go to the Wards page', async () => {
+    await wardAppPage.goto();
+  });
+
+  await test.step('And I click on the “Manage” button', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:23]:
```suggestion
  await test.step('And I click the “Manage” button to manage pending admission requests', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+  await test.step('And I search for the patient', async () => {
+    await page.getByPlaceholder('Search for a patient by name or identifier number').fill(patient.person.display);
+  });
+
+  await test.step('And I select the patient', async () => {
+    await page.getByRole('button', { name: patient.person.display }).click();
+  });
+
+  await test.step('And I click on the “Admit patient” button', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:39]:
```suggestion
  await test.step('And I click the “Admit patient” button', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  await test.step('And I click on the “Manage” button', async () => {
+    await page.getByRole('button', { name: 'Manage' }).click();
+  });
+
+  await test.step('And I click on the “Add Patient” button', async () => {
+    await page.getByLabel('Add patient to ward').click();
+  });
+
+  await test.step('And I search for the patient', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:31]:
```suggestion
  await test.step('And I fill in the patient name into the search box', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByPlaceholder('Search for a patient by name or identifier number').fill(patient.person.display);
+  });
+
+  await test.step('And I select the patient', async () => {
+    await page.getByRole('button', { name: patient.person.display }).click();
+  });
+
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:38]:
```suggestion

  await test.step('Then I should see the patient details show up in the "Create admission encounter" workspace', async () => {
     await page.getByRole('button', { name: 'Admit patient' }).click();
     await page.getByRole('button', { name: 'Back to search results' }).click();
  });

```

File: e2e/specs/wards-managment.spec.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  await test.step('And I click on the “Admit patient” button', async () => {
+    await page.getByRole('button', { name: 'Admit patient' }).click();
+  });
+
+  await test.step('And I click on the “Admit” button', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:43]:
```suggestion
  await test.step('Then I click the “Admit” button in the Admission requests workspace', async () => {
    await expect(page.getByText('Admission requests').toBeVisible();
```

File: e2e/specs/wards-managment.spec.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByRole('button', { name: 'Admit', exact: true }).click();
+  });
+
+  await test.step('Then I should see a success message', async () => {
+    await expect(page.getByText('Patient admitted successfully', { exact: true })).toBeVisible();
+  });
+
+  await test.step('And I should see the patient in the ward page', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:51]:
```suggestion
  await test.step('And I should see the patient's ward card added to the Ward page', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  });
+
+  await test.step('Then i select "Patient note" and add the note', async () => {
+    await page.getByTestId(patient.uuid).click();
+    await page.getByLabel('Patient note').click();
+    await page.getByPlaceholder('Write any notes here').fill('Test note');
+    await page.getByRole('button', { name: 'Save', exact: true }).click();
+  });
+
+  await test.step('And I should see a success message for successfully submitting the note', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:62]:
```suggestion
  await test.step('Then I should see a success message confirming the note was submitted', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    await expect(page.getByText(/patient note saved/i)).toBeVisible();
+  });
+
+  await test.step('Then I should transfer the patient to a different Ward', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:66]:
```suggestion
  await test.step('And when I transfer the patient to a different Ward', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByTestId(patient.uuid).click();
+    await page.getByLabel('Transfers').click();
+    await page.getByPlaceholder('Search locations').fill('Inpatient Ward');
+    await page.getByText('Inpatient Ward').click();
+    await page.getByRole('button', { name: 'Save', exact: true }).click();
+  });
+
+  await test.step('And I should see a success message for creating transfer request of the patient', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:74]:
```suggestion
  await test.step('Then I should see a success message confirming a transfer request was created for the patient', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+  });
+
+  await test.step('Then I should change the location to Inpatient', async () => {
+    await page.getByLabel('Change location').click();
+    await page.getByPlaceholder('Search for a location').fill('Inpatient Ward');
+    await page.getByText('Inpatient Ward').click();
+    await page.getByRole('button', { name: 'Confirm' }).click();
+  });
+
+  await test.step('And I should see a success message for changing the location of the patient', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:85]:
```suggestion
  await test.step('Then I should see a success message confirming the patient's location was updated', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    await expect(page.getByText(/Location updated/i)).toBeVisible();
+  });
+
+  await test.step('Then I should accept the patient admission request in the Inpatient Ward', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:89]:
```suggestion
  await test.step('And when I accept the pending admission request in the Inpatient Ward', async () => {
```

File: e2e/specs/wards-managment.spec.ts [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    await page.getByRole('button', { name: 'Manage' }).click();
+    await page.getByRole('button', { name: 'Transfer patient', exact: true }).click();
+  });
+
+  await test.step('And I should see the success message and the patient in the Inpatient Ward', async () => {
```
Comment (denniskigen) [inline e2e/specs/wards-managment.spec.ts:94]:
```suggestion
  await test.step('Then I should see a success message confirming the patient was admitted to the Inpatient Ward', async () => {
```

**Comment by jayasanka-sack** [comment]
Hey @jayg2002, this PR has been inactive for quite some time and the codebase has changed significantly since it was opened.

We’re marking it as graveyard-candidate and may close it soon as part of repository cleanup.

If you're still interested in continuing the work, please let us know and we'd be happy to help revive it.
